server/src/main/java/org/opensearch/monitor/os/OsProbe.java

```diff
@@ -413,8 +418,17 @@ String readSysFsCgroupCpuAcctCpuAcctCfsQuota(final String controlGroup) throws I
                 case "throttled_time":
                     timeThrottledNanos = Long.parseLong(fields[1]);
                     break;
+                default:
+                    unknownFields.add(fields[0]);
+                    break;
             }
         }
+        if (unknownFields.isEmpty() == false) {
+            logger.warn(
+                "Expected to see only nr_periods, nr_throttled, throttled_time values, but also got {} values",
+                String.join(", ", unknownFields)
```

Comment:
> 1. Who calls this code? Are we going to get flooded by warnings?

Double checked the code. `OsProde` is called by `OsService` each second by default (system property `monitor.os.refresh_interval`) so such implementation will definitely flood by warnings messages.

Here is the code:
https://github.com/opensearch-project/OpenSearch/blob/f0984eb409d44e8b68deb1c262bf81accc300acb/server/src/main/java/org/opensearch/monitor/os/OsService.java#L62

And based on the git blame, the severity for some methods in this class was changes from warn to debug in 2020, so it looks like that the warn level is not the best idea. Better to use - debug.   

> 2. Are unknown fields really a problem or should we just ignore that? I was thinking that if one of the stats is missing (e.g. `throttled_time`) we won't know it from the warn message. 

Good question, from the one hand it is kind of a best practice to use logging for unknown fields, so somebody could add them, but on the other hand if nobody needs them it could be annoying. At our company we do not use JMX statistics provided by Opensearch for its process.

> Warn if a) any field is missing,  b) any field value is -1. Additional fields seem ok, but I could be wrong. I do think it would be nice to know in the message what was returned in case there's output that we didn't expect.
-1 values check later in the code using `assert`:
https://github.com/opensearch-project/OpenSearch/blob/f0984eb409d44e8b68deb1c262bf81accc300acb/server/src/main/java/org/opensearch/monitor/os/OsProbe.java#L418

Overall:
I think that we can ignore logging of unknown fields, 
but add a debug message for missing fields or fields with -1 value. 
TBH I can not imaging a situation when `cgroups` do not provide CPU stats (at least I did not find any info about default values for this fields), only if kernel in PANIC...   
What do you think?

Reply:
I like everything you said. 

We do have a mechanism to warning just once which is probably what we can/should do here regardless of the message severity, see https://github.com/opensearch-project/OpenSearch/issues/1108 for an example.

Reply:
Regarding the unknown fields, my two cents is to not log unknown fields. I'd bias towards keeping it simple unless we have a really good reason to change it. #1108 involves warning about usage of deprecated features (which is important to log) versus here where unknown fields may or may not be useful to know about.

Reply:
I suppose a debug log may be ok here, logs is one way to notice there are new CPU features and someone thinking about adding support for them. If this code just had `>= 3` already we'd never know.

No strong feelings either way.

Reply:
OK wont log unknown fields. Better to add them on-demand using features. If you do not mind I will remove all `assert` IMHO it is a bad practice to use such things it could freeze development process, since assert throws `NullPointerException`. The idea is to replace them with warning messages instead which will be logged only once by adding a boolean flag or bits flag without synchronization since `OsService` creates only once during boot-up of a node. WDYT?

Reply:
`assert` will throw an AssertionError right?

I'm personally not a big fan of using the assert keyword, but it seems to be widely used within OpenSearch. The only problem with removing them in a case like this is that something downstream may fail unpredictably if it encounters a `-1` value whereas currently it fails immediately here (assuming assertions are enabled via the JVM argument). If that seems like a minimal riskthen I have no problem replacing the assertions with your suggestion.

Reply:
> `assert` will throw an AssertionError right?
Yes true. Mixed it up exception looks like this:
```bash
» ERROR][o.o.b.OpenSearchUncaughtExceptionHandler] [integTest-0] fatal error in thread [main], exiting
»  java.lang.AssertionError: null
»       at org.opensearch.monitor.os.OsProbe.readSysFsCgroupCpuAcctCpuStat(OsProbe.java:443) ~[opensearch-1.2.4-SNAPSHOT.jar:1.2.4-SNAPSHOT]
»       at org.opensearch.monitor.os.OsProbe.getCgroupCpuAcctCpuStat(OsProbe.java:400) ~[opensearch-1.2.4-SNAPSHOT.jar:1.2.4-SNAPSHOT]
»       at org.opensearch.monitor.os.OsProbe.getCgroup(OsProbe.java:555) ~[opensearch-1.2.4-SNAPSHOT.jar:1.2.4-SNAPSHOT]
»       at org.opensearch.monitor.os.OsProbe.osStats(OsProbe.java:676) ~[opensearch-1.2.4-SNAPSHOT.jar:1.2.4-SNAPSHOT]
»       at org.opensearch.monitor.os.OsService.<init>(OsService.java:66) ~[opensearch-1.2.4-SNAPSHOT.jar:1.2.4-SNAPSHOT]
»       at org.opensearch.monitor.MonitorService.<init>(MonitorService.java:57) ~[opensearch-1.2.4-SNAPSHOT.jar:1.2.4-SNAPSHOT]
»       at org.opensearch.node.Node.<init>(Node.java:502) ~[opensearch-1.2.4-SNAPSHOT.jar:1.2.4-SNAPSHOT]
»       at org.opensearch.node.Node.<init>(Node.java:319) ~[opensearch-1.2.4-SNAPSHOT.jar:1.2.4-SNAPSHOT]
»       at org.opensearch.bootstrap.Bootstrap$5.<init>(Bootstrap.java:242) ~[opensearch-1.2.4-SNAPSHOT.jar:1.2.4-SNAPSHOT]
»       at org.opensearch.bootstrap.Bootstrap.setup(Bootstrap.java:242) ~[opensearch-1.2.4-SNAPSHOT.jar:1.2.4-SNAPSHOT]
»       at org.opensearch.bootstrap.Bootstrap.init(Bootstrap.java:412) ~[opensearch-1.2.4-SNAPSHOT.jar:1.2.4-SNAPSHOT]
»       at org.opensearch.bootstrap.OpenSearch.init(OpenSearch.java:178) ~[opensearch-1.2.4-SNAPSHOT.jar:1.2.4-SNAPSHOT]
»       at org.opensearch.bootstrap.OpenSearch.execute(OpenSearch.java:169) ~[opensearch-1.2.4-SNAPSHOT.jar:1.2.4-SNAPSHOT]
»       at org.opensearch.cli.EnvironmentAwareCommand.execute(EnvironmentAwareCommand.java:100) ~[opensearch-1.2.4-SNAPSHOT.jar:1.2.4-SNAPSHOT]
»       at org.opensearch.cli.Command.mainWithoutErrorHandling(Command.java:138) ~[opensearch-cli-1.2.4-SNAPSHOT.jar:1.2.4-SNAPSHOT]
»       at org.opensearch.cli.Command.main(Command.java:101) ~[opensearch-cli-1.2.4-SNAPSHOT.jar:1.2.4-SNAPSHOT]
»       at org.opensearch.bootstrap.OpenSearch.main(OpenSearch.java:135) ~[opensearch-1.2.4-SNAPSHOT.jar:1.2.4-SNAPSHOT]
»       at org.opensearch.bootstrap.OpenSearch.main(OpenSearch.java:101) ~[opensearch-1.2.4-SNAPSHOT.jar:1.2.4-SNAPSHOT]
```
Just saw `null` and automatically decided that it is `NullPointerException` :-) 

> 
> I'm personally not a big fan of using the assert keyword, but it seems to be widely used within OpenSearch. The only problem with removing them in a case like this is that something downstream may fail unpredictably if it encounters a `-1` value whereas currently it fails immediately here (assuming assertions are enabled via the JVM argument). If that seems like a minimal riskthen I have no problem replacing the assertions with your suggestion.

Then in this case better to add this functionality so far for CPU stat only and leave other places as is.